ethcore/src/client/test_client.rs

```diff
@@ -351,17 +370,116 @@ impl TestBlockChainClient {
+			options: TransactOptions<T, V>,
+		) -> Result<Executed<T::Output, V::Output>, CallError> where
+			T: trace::Tracer,
+			V: trace::VMTracer,
+		{
+			let options = options
+				.dont_check_nonce()
+				.save_output_from_contract();
+			let original_state = if state_diff { Some(state.clone()) } else { None };
+
+			let mut ret = Executive::new(state, env_info, machine).transact_virtual(transaction, options)?;
+
+			if let Some(original) = original_state {
+				ret.state_diff = Some(state.diff_from(original).map_err(ExecutionError::from)?);
+			}
+			Ok(ret)
+		}
+
+		let state_diff = analytics.state_diffing;
+		let machine = self.spec.engine.machine();
+
+		match (analytics.vm_tracing,) {
+			(true,) => call(state, env_info, machine, state_diff, t, TransactOptions::with_vm_tracing()),
+			(false,) => call(state, env_info, machine, state_diff, t, TransactOptions::with_no_tracing()),
+		}
+	}
+
+    /// Convenience method to retrieve current state from a given state_db
+    pub fn get_in_memory_state(&self) -> Result<State<StateDB>, EvmTestError> {
+        {
+	        let state_db = self.state_db.lock().boxed_clone();
+
+    	    State::from_existing(
+    	    	state_db,
+    	    	*self.spec.genesis_header().state_root(),
+    	    	self.spec.engine.account_start_nonce(0),
+    	    	factories(),
+    	    ).map_err(EvmTestError::Trie)
+        }
+    }
+    
+    /// Convenience method to retrieve state from a fresh state_db
+    pub fn get_new_state(&self) -> Result<State<StateDB>, EvmTestError> {
+        let (_, state_db) = get_in_memory_db(&self.spec).unwrap();
+    	State::from_existing(
+    		state_db.into_inner(),
+    		*self.spec.genesis_header().state_root(),
+    		self.spec.engine.account_start_nonce(0),
+    		factories(),
+    	).map_err(EvmTestError::Trie)
+    }
+}
+
+pub fn factories() -> Factories {
+    Factories { 
+        vm: EvmFactory::new(VMType::Interpreter, 1024 * 1024), 
+        trie: trie::TrieFactory::new(trie::TrieSpec::Secure), 
+        accountdb: Default::default(), 
+    } 
 }
 
-pub fn get_temp_state_db() -> GuardedTempResult<StateDB> {
-	let temp = RandomTempPath::new();
-	let db = Database::open(&DatabaseConfig::with_columns(NUM_COLUMNS), temp.as_str()).unwrap();
-	let journal_db = journaldb::new(Arc::new(db), journaldb::Algorithm::EarlyMerge, COL_STATE);
-	let state_db = StateDB::new(journal_db, 1024 * 1024);
-	GuardedTempResult {
-		_temp: temp,
-		result: Some(state_db)
+pub fn get_in_memory_db(spec: &Spec) -> Result<(RwLock<Arc<KeyValueDB>>, Mutex<StateDB>), EvmTestError> {
+	let db = Arc::new(kvdb_memorydb::create(NUM_COLUMNS.expect("We use column-based DB; qed")));
+	let journal_db = journaldb::new(db.clone(), journaldb::Algorithm::EarlyMerge, COL_STATE);
+	let mut state_db = StateDB::new(journal_db, 5 * 1024 * 1024);
+
+	let genesis = spec.genesis_header();
+	// Write DB
+	{
+		let mut batch = kvdb::DBTransaction::new();
+		state_db.journal_under(&mut batch, 0, &genesis.hash())?;
```

Comment:
`state_db` is empty at this point, no? I don't think there is a need to write anything to the DB.

Reply:
Unsure, I took this from the existing code, and modified it to use the new `state_db` setup.

> I don't think there is a need to write anything to the DB.

Why wouldn't we need to add the genesis hash to the DB when it is created?